flink-tests/src/test/java/org/apache/flink/test/checkpointing/RestoreUpgradedJobITCase.java

```diff
@@ -0,0 +1,419 @@
+import static org.hamcrest.Matchers.is;
+
+/**
+ * Test check scenario when the upgraded job(different map order, different record type, new map)
+ * restored on old savepoint/checkpoint.
+ */
+@RunWith(Parameterized.class)
+public class RestoreUpgradedJobITCase extends TestLogger {
+    private static final int PARALLELISM = 3;
+    private static final int TOTAL_RECORDS = 100;
+
+    @ClassRule public static TemporaryFolder temporaryFolder = new TemporaryFolder();
+    @Parameterized.Parameter public TestCheckpointType checkpointType;
+
+    @ClassRule
+    public static final MiniClusterWithClientResource CLUSTER =
+            new MiniClusterWithClientResource(
+                    new MiniClusterResourceConfiguration.Builder()
+                            .setConfiguration(new Configuration())
+                            .setNumberTaskManagers(2)
+                            .setNumberSlotsPerTaskManager(10)
+                            .build());
+
+    @Rule public final SharedObjects sharedObjects = SharedObjects.create();
+
+    private SharedReference<OneShotLatch> allDataEmittedLatch;
+    private SharedReference<AtomicLong> result;
+
+    public void setupSharedObjects() {
+        allDataEmittedLatch = sharedObjects.add(new OneShotLatch());
+        result = sharedObjects.add(new AtomicLong());
+    }
+
+    @Parameterized.Parameters(name = "Savepoint type[{0}]")
+    public static Object[][] parameters() {
+        return new Object[][] {
+            {ALIGNED_CHECKPOINT}, {CANONICAL_SAVEPOINT}, {NATIVE_SAVEPOINT},
+        };
+    }
+
+    enum TestCheckpointType {
+        ALIGNED_CHECKPOINT,
+        CANONICAL_SAVEPOINT,
+        NATIVE_SAVEPOINT
+    }
+
+    enum MapName {
+        MAP_1,
+        MAP_2,
+        MAP_3,
+        MAP_4,
+        MAP_5,
+        MAP_6;
+
+        int id() {
+            return ordinal() + 1;
+        }
+    }
+
+    @Test
+    public void testRestoreUpgradedJob() throws Exception {
+        setupSharedObjects();
+
+        // given: Stream with all network exchanges.
+        StreamExecutionEnvironment env = originalStream();
+
+        // when: Job is executed.
+        JobClient jobClient = env.executeAsync("Total sum");
+        waitForAllTaskRunning(CLUSTER.getMiniCluster(), jobClient.getJobID(), false);
+
+        allDataEmittedLatch.get().await();
+        allDataEmittedLatch.get().reset();
+
+        String savepointPath = stopWithCheckpointOrSavepoint(jobClient);
+
+        // then: Check the result before the checkpoint.
+        assertThat(result.get().longValue(), is(calculateExpectedResultBeforeSavepoint()));
+        result.get().set(0);
+
+        // when: Executing the new job with different order of maps.
+        env = upgradedStream(savepointPath);
+        jobClient = env.executeAsync("Total sum");
+
+        waitForAllTaskRunning(CLUSTER.getMiniCluster(), jobClient.getJobID(), false);
+
+        allDataEmittedLatch.get().await();
```

Comment:
Yes, we need both of them since the source can emit all data before all other subtasks would be ready and vice versa if all tasks are ready it doesn't mean that the source has emitted all results already.